apache/zeppelin
# [ZEPPELIN-1850] Introduce Webpack (front)

### What is this PR for?

<br/>

**1. The purpose of this PR is introducing webpack to zeppelin-web**

- It supports super-fast javascript compilation and (hot) reload
- It helps to simplify, unify build process.
- It enables for us to use `import`, `export` features.
- It's more modern stack than grunt. This will encourage other developers to contribute (*the most import thing IMO*).

**2. This PR is designed to improve build process _gradually!!!_**

- So, Angular module loading depends on the import sequence as we did before in [index.html](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/index.html#L171). (see [index.js](https://github.com/apache/zeppelin/pull/1805/files#diff-a128f4667c7e1afb6058389c20627b4e)). 
- Also, CSS and HTML file is not managed by webpack including its live reloading

*These will be handled by additional PRs.*

**3. This PR is not big**

- 60+ file changes are just about removing `use strict`, importing `zeppelin.js`. (Also includes fixups for #1802 and #1803)
- So skip them while reveiw
- Please focus on the changes in `Gruntfile.js`, `package.json`, `webpack.config.js`

<br/>

### What type of PR is it?
[Improvement]

### Todos
* [x] - Setup webpack.config.js
* [x] - Resolve global variable `zeppelin` problem
* [x] - Support webpack in karma (`npm run test`)
* [x] - Annotate the result bundle using webpack ng annotate plugin
* [x] - Fix eslint violations
* [x] - Livereload for HTML, CSS

### What is the Jira issue?

[ZEPPELIN-1850](https://issues.apache.org/jira/browse/ZEPPELIN-1850)

### How should this be tested?

- `cd zeppelin-web && rm -rf node_modules bower_components node`
- `npm install`
- `npm run test`
- `npm run build`
- `npm run dev` and open `localhost:9000`: **check live-reload works regarding to html, css, js files**
- `cd .. && mvn clean package -pl 'zeppelin-web' -DskipTests &&  ./bin/zeppelin-daemon.sh restart` and open `localhost:8080`

### Screenshots (if appropriate)

N/A

### Questions:
* Does the licenses files need update? - NO
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - NO
